### ISSUE TITLE
Bash completion for `start` is very slow

### ISSUE THREAD

**Issue by danbst** [systemd/systemd#7185]
### Submission type

 - Request for enhancement (RFE)

### systemd version the issue has been seen with

> systemd v234

### Used distribution

> NixOS 17.09

Bash completion works fine for `systemctl` subcomands, except it is very slow for some of them. For example, it is less then 1 second for `stop`, `kill` subcommands, but > 15 sec for `start`, `restart`, `status` subcommands.

I have 227 units in `systemctl -a | wc -l`

There is a patch hanging around (https://www.reddit.com/r/systemd/comments/5ze9jz/systemd_systemctl_bash_completion_is_very_slow_or/), but I haven't yet tested it

**Comment by keszybz** [comment]
A PR would be appreciated (and easier to discuss).

**Comment by danbst** [comment]
Indeed, I tried that patch and got no speedup. But I've done some debugging also.

`systemctl stop` and `systemctl start` trigger different functions in bash completion code. The function which handles `start` is called `__filter_units_by_properties` and calls internally `systemctl list-unit-files`. And it makes a real slowdown here!

```
$ time systemctl --no-pager list-unit-files >/dev/null

real: 0m9.467s
```
However `list-units` works instantly.

Also note, I'm running these tests in a QEMU VM, which adds another performance penalty.

**Comment by danbst** [comment]
Also, found related issue https://github.com/systemd/systemd/pull/6927, probably I can prepare something similar for bash

**Comment by keszybz** [comment]
For `stop` it is OK to just look at units which are already in memory. But for `start` this is not an option, actually people usually want to start units which haven't been started yet, so units which often aren't loaded.

**Comment by aphirst** [comment]
I was curious about this too. There doesn't seem to be a (bash-related) PR for this, and were I not entirely unfamiliar with the codebase, I might have been willing to submit one myself. Since the ZSH equivalent(?) was merged, is there a reasonable chance that this could be too?

**Comment by evelikov** [comment]
Everyone feel free to test the above series. One could polish even more, but as-is it makes a notable improvement.

It seems that the execution time of systemctl commands, depends on the size of the journal [1].
I'm not 100% certain about this, so take it with a healthy grain of salt.

[1] http://www.codekoala.com/posts/systemctl-status-foo-was-too-slow/

**Comment by sammcj** [comment]
- Patch from: [UHUsystems.com] (https://uhusystems.com/)
- Maintainer: Janos Strbik janos.strbik@uhusystems.com

---

- Problem: bash completion of systemctl command is very slow or totally hanged up for minutes in case of a great number of systemd units (>250).
- Reason: Completion routines calls systemctl as times as many systemd units exist.
- Solution: Patch file `/usr/share/bash-completion/completions/systemctl` as the following:

```patch
22c22
<         systemctl $mode --full --no-legend "$@"
---
>         systemctl $mode --full --no-legend --no-pager "$@"
44a45,48
> 
> #
> # UHU PATCH Tue Mar 14 18:50:31 CET 2017 {
> #
46,49c50,53
<             <(__systemctl $mode show --property "$property" -- "${units[@]}")
<         for ((i=0; $i < ${#units[*]}; i++)); do
<                 if [[ "${props[i]}" = "$property=$value" ]]; then
<                         echo " ${units[i]}"
---
>             <(__systemctl $mode show --property Names,"$property" -- \*)
>         for ((i=0; $i < ${#props[*]}; i=i+2)); do
>                 if [[ "${props[i+1]}" = "$property=$value" ]]; then
>                         echo " ${props[i]#Names=}"
51a56,59
> #
> # } UHU PATCH Tue Mar 14 18:50:31 CET 2017
> #
> 
```

Source: https://www.reddit.com/r/systemd/comments/5ze9jz/systemd_systemctl_bash_completion_is_very_slow_or/

**Comment by yuwata** [comment]
I've submitted f28255e2d5695b0d060558ff8cbb2b526fc94cc0. Is it still slow? If so, please try to submit revised one.

**Comment by yuwata** [comment]
> in case of a great number of systemd units (>250)

In such a situation, hitting `systemctl start <TAB>` is not a good idea. And when you type the first few characters, e.g. `systemctl start syst<TAB>`, I hope the performance is drastically improved now.

**Comment by elico** [comment]
any patches for Ubuntu 18.04?

**Comment by mrc0mmand** [comment]
Well, the bash completion's performance for `systemctl` has been _significantly_ improved thanks to @yuwata, so I'd suggest trying the latest upstream version of bash completions (https://github.com/systemd/systemd/tree/master/shell-completion/bash) and if they work for you, request a backport from your distribution.

**Comment by elico** [comment]
@mrc0mmand I have tried to patch the 18.04 bash completion manually but there is a big difference between the current version to the distribution one.
It's too much for me to patch also since I am not too familiar with these parts of the code.
On CentOS 7 the default is slow but acceptable.

Also I wanted to respond that the patch in @sammcj may have helped if I have few units with a specific prefix but for a simple `systemctl start <TAB>` I think that nothing will help.

@mrc0mmand on what OS/distribution I can test the latest patches? I was thinking about ArchLinux.

**Comment by mrc0mmand** [comment]
@elico in case of `systemctl` bash completions you could, theoretically, do something like this:

```bash
wget -O systemctl https://raw.githubusercontent.com/systemd/systemd/master/shell-completion/bash/systemctl.in
sed -i "s#@rootlibexecdir@#$(pkg-config --variable systemdutildir systemd)#" systemctl
source ./systemctl
```

This should allow you to temporarily override the shipped bash completions for `systemctl` with the upstream ones.

> @mrc0mmand on what OS/distribution I can test the latest patches? I was thinking about ArchLinux.

Arch Linux or Fedora should suffice.

**Comment by AddisonG** [comment]
This is still a huge issue - accidentally hitting <tab> 3 or 4 times makes the entire terminal hang for 20 seconds. The code fix from reddit doesn't seem to improve anything

**Comment by yuwata** [comment]
@AddisonG Which version of systemd do you use?

**Comment by mvvvmd** [comment]
I can confirm this issue, systemctl consistently hangs on tab completion or fails to complete all together. This is on all supported Ubuntu LTS versions. The highest systemd version we run is 237-3. 

Tab completion in applications other than systemd works fast with small delays on slower storage and large directories for instance. 

It would be great if this can be fixed as it is a daily disruption for admins and it has been for years.

**Comment by mrc0mmand** [comment]
> I can confirm this issue, systemctl consistently hangs on tab completion or fails to complete all together. This is on all supported Ubuntu LTS versions. The highest systemd version we run is 237-3.

Here's the problem - the performance fix (along with several other patches) were introduced in systemd v240 and later. Could you try the temporary solution from https://github.com/systemd/systemd/issues/7185#issuecomment-495984238 to see if the issue is still present even with the latest bash completions? If not, you'll have ask the systemd maintainers in Ubuntu to backport relevant patches.

**Comment by mrc0mmand** [comment]
Also, as someone pointed out on @yuwata's commit, it indeed fixes the issue, but it has not been backported yet: https://github.com/systemd/systemd/commit/f28255e2d5695b0d060558ff8cbb2b526fc94cc0#commitcomment-33845642

**Comment by mvvvmd** [comment]
Thank you for the information, it wasn't clear to me the fix made it to a systemd release. I've tested the latest version, it seems fixed like you said. 

Best to close this ticket I think, although people will likely keep reporting as it seems at least Ubuntu do not or are slow to backport.

**Comment by yuwata** [comment]
OK, let's close this.

**Comment by colthreepv** [comment]
anyone coming here for ubuntu 18.04 can use this PPA from @ddstreet 
https://launchpad.net/~ddstreet/+archive/ubuntu/systemd-upstream

**Comment by ddstreet** [comment]
@colthreepv while anyone's welcome to use my ppa, i wouldn't suggest using it for anything important.  If this should be fixed in Ubuntu can you open a bug in launchpad?  I can't promise it will get fixed in Ubuntu 18.04 but we can at least track/review it for Ubuntu with a LP bug.

**Comment by mvvvmd** [comment]
It would save us 2+ years of annoyance for sure, there is a bug report on launchpad already here:

https://bugs.launchpad.net/ubuntu/+source/bash-completion/+bug/1529152
